File: leetcode/Hot-100/dcecs.py
"""
给定一个二叉树，检查它是否是镜像对称的。
例如：
    1
   / \
  2   2
 / \ / \
3  4 4  3
"""
class Solution:
    def isSymmetric(self, root):
        if not root:
            return True
        return self.dfs(root.left, root.right)    
    def dfs(self, left, right):
        if not left and not right:
            return True
        if not left or not right:
            return False
        return (left.val == right.val) and self.dfs(left.left, right.right) and self.dfs(left.right, right.left)

File: leetcode/Hot-100/c++/dcecs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool dfs(TreeNode* left, TreeNode* right){
        if (!left && !right) {
            return true;
        } 

        if (!left || !right) {
            return false;
        }

        return (left->val == right->val && dfs(left->left, right->right) && dfs(left->right, right->left));
    }

    bool isSymmetric(TreeNode* root) {
        if (!root) {
            return true;
        }
        return dfs(root->left, root->right);
    }

    
};